src/ApplicationInsights.Kubernetes/TelemetryInitializers/TelemetryKeyCache.cs

```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Collections.Concurrent;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.ApplicationInsights.Kubernetes
+{
+    internal class TelemetryKeyCache : ITelemetryKeyCache
+    {
+        // The capacity of 40 is kind of arbitrary. We have around 10 keys appended to for the telemetry;
+        // Times that by 4 to allow a bit of wiggling room for name mapping;
+        // However, having too many of the mapped key is not allowed to prevent huge memory consuption.
+        internal const int CacheCapacity = 40;
+        private readonly AppInsightsForKubernetesOptions _options;
+
+        private ConcurrentDictionary<string, string> _telemetryKeyCache;
+
+        public TelemetryKeyCache(IOptions<AppInsightsForKubernetesOptions> options)
+        {
+            _options = options?.Value ?? throw new System.ArgumentNullException(nameof(options));
+
+            if (_options.TelemetryKeyProcessor != null)
+            {
+                _telemetryKeyCache = new ConcurrentDictionary<string, string>(
+                    concurrencyLevel: 16,
```

Comment:
Probably not necessary. According to https://devblogs.microsoft.com/pfxteam/concurrentdictionary-performance-improvements-in-net-4-5/ the ConcurrentDictionary will dynamically change the number of locks used if concurrencyLevel is not specified.